### ISSUE TITLE
Can this be used in php ?

### ISSUE THREAD

**Issue by kressly** [mozilla/mozjpeg#345]
Can this be used in php ?

**Comment by kornelski** [comment]
Yes. There are two ways of doing it. 

* The good and efficient way is to compile PHP with MozJPEG instead of libjpeg. When compiling it specify mozjpeg's location in `--with-jpeg-dir`.

* The poor way is to use `cjpeg` demo tool and feed it image files, as explained here: https://calendar.perfplanet.com/2014/mozjpeg-3-0/

**Comment by exikyut** [comment]
I sadly don't think there's currently a way to pass through a scan script when using the (much) more efficient link-with-MozJPEG approach.

Scan scripts appear to have originated with libjpeg-turbo (unsure if they existed in original libjpeg).

PHP's main image manipulation library, libgd, doesn't provide any way to plumb this sort of functionality through to scripts.

Imagick is relatively straightforward to install (via package management) but I don't think it provides a passthrough either.

Making a libjpeg-turbo PHP extension wouldn't solve this, because it would have to become Yet Another Image Manipulation Library in the process of exposing all of libjpeg's capabilities to scripts in a useful way.

Others with actual experience might be able to constructively comment more usefully here, but I personally suspect that extending libgd and/or ImageMagick/GraphicsMagick to provide these kinds of capabilities would also suggest fine control over other aspects of image development, which these libraries don't generally do (particularly GD).

TL;DR: What sounds the most viable to me is a PHP extension that can be handed PNG/PPM/JPEG images and then express extremely fine control over the encoding stages. PHP also has a reputation for "importing" objects between extensions where it makes sense to do so (eg, socket->stream, DOMDocument<->SimpleXML), so maybe the extension could "import" GD objects directly.

I personally reckon it would definitely be worth it to make the effort to figure out how to make this work, so I filed https://github.com/libgd/libgd/issues/663 to put this on their radar at a high level, in case there is interest.

**Comment by pierrejoye** [comment]
Also, adding a comment here too, for the record.

I am not sure we can have a generic approach in GD. However, most of these specific features like custom quantization tables, sampling factors etc, could be easily made available based on which JPEG lib is used (or any other specific format), could be made available using specific APIs, basically exposing them to the user (be the ext/gd in php or any other libgd user). After that, advanced users can handle them when desired.

At the libgd level, we try to keep it simple for the common users. One without knowledge of these specifics may not want to deal with that and PHP, or libgd, is also not knowledgeable enough to start to mess with these parameters (the author of the jpeg libs know better and we rely on them for the default).

**Comment by kornelski** [comment]
MozJPEG enables its best features by default, so APIs for selecting quantization tables, etc. aren't strictly necessary. Just a drop-in replacement should be sufficient.

**Comment by ygoe** [comment]
Could I just install mozjpeg as an Ubuntu package and it would be used instead of the default library? Or is it linked statically and doesn't care about installed libraries? I'm talking about standard PHP with GD. Actually I'm using these here: `add-apt-repository ppa:ondrej/php`

**Comment by AGenchev** [comment]
Does it retain the EXIF data in the recompressed files like the Imagick does ?

**Comment by kornelski** [comment]
The `cjpeg` demo tool has `-copy all` option.

**Comment by ampaze** [comment]
For future reference, you don't necessarily need to recompile PHP to be able to use mozjpeg. In Linux you can use the LD_PRELOAD environment variable like so:
```bash
LD_PRELOAD=/path/to/mozjpeg/lib/libjpeg.so
```
I got it working with php-fpm (using gd) and supervisord with this method.
